### ISSUE TITLE
duk_def_prop() -like API call whose argument is a property descriptor object

### ISSUE THREAD

**Issue by svaarala** [svaarala/duktape#1088]
Using an object argument is more easily extensible, and it's quite possible custom Duktape-specific property attributes are needed over time. See discussion in #759 re: API naming options.

**Comment by fatcerberus** [comment]
Chakra's [tt]JsDefineProperty()[/tt] already works like this:
https://docs.microsoft.com/en-us/scripting/chakra-hosting/jsdefineproperty-function

So I guess it wouldn't be so weird to have Duktape work similarly after all. 😄

**Comment by svaarala** [comment]
I don't see how that changes the arguments for or against, but glad you agree :)

**Comment by fatcerberus** [comment]
If you remember, I thought it was kind of awkward to use an Ecmascript object for a C API in this context.  Now that I see there's precedent for it elsewhere I'm willing to concede that maybe it's not so bad :)

**Comment by svaarala** [comment]
There's also precedent in a many non-language-related C libraries that implement some form of key/value property sets for APIs that need continuous revision and extension. Sometimes it's an explicit key/value holder (like an object) while sometimes there's a setProperty(key,value) call that is called on some target (e.g. Curl uses the latter).

**Comment by svaarala** [comment]
I think that the key/value holder or setProperty(key,value) pattern makes sense even for pure C libraries in cases where a lot of parameters are needed, changes and extensions are anticipated, and performance is not critical. Dealing with such an API scenario in a pure C fashion is otherwise a bit painful: one either needs a lot of functions, a lot of arguments for function(s), packing arguments into C structs (which are difficult to version), struct versioning or macro initializers, etc.

For an Ecmascript engine it's just a nice bonus that the language object type can be reused for this purpose. It could, for most purposes, equally well be a separate key/value holder type but the gains from doing so seem minimal and probably not worth it; there might be some RAM benefit from doing so, but footprint would increase to support that additional type.

However, using a separate key/value holder for parameters would actually allow one to solve one downside of the general argument object pattern: with just an Ecmascript object parameter the caller cannot signal which keys are critical (= reject operation if not recognized) and which are not (= ignore if not recognized). With a specific key/value parameter holder a criticality flag can be a part of each key/value. But, I don't think solving this issue is worth adding another concept to the API, unless there's a very clear use case where this is really important. Having a notion of criticality also adds one more thing to understand when using the API.

A really problematic API scenario is one where extensions are anticipated and the call is performance critical. I don't know of a good C API pattern to solve that. Maybe closest is an undisclosed struct with macro initializers which can be extended, which versions reasonably well but is a bit messy in the call site. Luckily there aren't such cases in the API right now I think.

Finally, for this specific API call, duk_def_prop(), using an object is also nice because it matches how the Ecmascript counterpart works. Although it's maybe a bit weird if Object.defineProperty() doesn't accept all the same custom attributes as duk_def_prop(). Also some arguments like a force flag (DUK_DEFPROP_FORCE now) can be included in the property descriptor but are not really part of the descriptor itself but rather additional flags related to the operation itself... so technically one might want two objects (a descriptor and an options object) but that seems a bit of an overkill also.

**Comment by svaarala** [comment]
> If you remember, I thought it was kind of awkward to use an Ecmascript object for a C API in this context. Now that I see there's precedent for it elsewhere I'm willing to concede that maybe it's not so bad :)

Fair enough :)

Though, it might actually be that an object is used for that API call because of the Object.defineProperty() counterpart, not because it was considered a good API design choice in general.

I would argue for the latter, for cases where extensibility is obviously needed and performance is not critical. The most obvious current candidates are duk_def_prop() and duk_push_proxy(). duk_debugger_attach() would also benefit from this, but there's no clean representation for the native function pointers of the callbacks :-/ So maybe that should adopt a C struct approach instead when the API is eventually revised.

**Comment by fatcerberus** [comment]
I think it's mainly the (perceived) GC overhead that bothers me about the proposal.  The key/value pair concept is sound and I definitely understand the benefits for versioning, it's just that I don't like the idea of generating extra garbage for an operation that could be handled in pure C (except for the actual value being assigned, of course).

I guess like you say though, it's not a performance-critical operation.

**Comment by svaarala** [comment]
In almost all cases a parameter object will be immediately collected via refcounting, so it's a very temporary allocation only. In theory the temporary allocation might increase memory fragmentation if the underlying allocator doesn't implement pools for small allocations (which almost always suffice for parameter objects). But in practice most allocators use pools to reduce fragmentation so it's mostly a non-issue.

**Comment by svaarala** [comment]
But I agree it's certainly a trade-off between versioning friendliness, performance, and footprint. It only really makes sense when performance and footprint are not especially critical. Footprint-wise the temporary objects are collected quickly as noted above, but sometimes even that temporary usage may matter, e.g. the allocator handles fragmentation poorly, or refcounting is disabled.

It's also sometimes difficult to be certain that an API call is not performance critical. For example, there are most likely (uncommon) cases where duk_def_prop() is part of some performance critical application code path. So ultimately it's a matter of judgement whether such uncommon cases are more important than flexible versioning.

The main merit of more flexible versioning for me is that it saves development time for both Duktape (less backwards compatibility management, less documentation for API call variants) and application code (fewer API deprecations etc).

**Comment by svaarala** [comment]
One more note about memory behavior: usually duk_def_prop() is a leaf call and doesn't nest so the memory usage is temporary. Two cases come immediately into mind where nesting may happen though:
- If mark-and-sweep happens during duk_def_prop(), finalizers may be executed which recursively call into duk_def_prop(). This shouldn't nest further though.
- If the target object is a Proxy with a `defineProperty` trap (not yet supported) the argument object gets interpreted, a standards-defined property descriptor object gets built, and the trap gets called. Depending on the implementation, the argument object can be freed without the proxy trap call. The trap itself may then nest arbitrarily.

So it's not a clear cut case but still in almost all normal cases the memory usage bump is only very temporary.

**Comment by fatcerberus** [comment]
In JSAL I made my def_prop work like the proposal here:
```c
jsal_push_eval("({ enumerable: false, writable: false, configurable: true })");
jsal_push_number(value);
jsal_set_named_property(-2, "value");
jsal_def_named_property(-2, name);
```

And I will say that's *much* less awkward than the flags.

**Comment by svaarala** [comment]
Does that first do a normal property write followed by setting of the attributes? If so, that's different from a straight Object.defineProperty() if the property already exists.

**Comment by fatcerberus** [comment]
No, it calls `Object.defineProperty` (or the C equivalent).  The first three lines are just constructing the property descriptor.

**Comment by svaarala** [comment]
Ok, I see now. That's pretty much what an object based duk_def_prop() would do, but:
- Needing an eval is an overkill for most call sites so IMO there must be an API call to define an object and multiple objects more efficiently.
- IMO the API to construct an object with properties must be able to pull values from the value stack (or index values on the value stack, more generically) rather than needing separate calls to set those properties.

**Comment by svaarala** [comment]
Side note: it would be nice if eval could somehow detect that something less than a full compilation could be used; e.g. in your example the initial descriptor could be parsed basically as a JSON object. This would allow call sites to use specific eval idioms more efficiently and they didn't need to worry about the difference between manual object construction, eval, decoding from a JSON string, etc.

Unfortunately detecting something like that seems quite impractical at run time.
